### PR TITLE
Sign any `x-amz-*` header

### LINKED ISSUE #27140: Sign aws specific headers `x-amz`
In https://github.com/ClickHouse/ClickHouse/pull/11134 support for custom headers in s3 functions was added, but it's not working if you would try to pass some aws specific headers like `x-amz-request-payer` in it.

```
<header>x-amz-request-payer: requester</header>
```
Error message:

```
`AccessDenied`, message: 'There were headers present in the request which were not signed'.
```

https://docs.aws.amazon.com/AmazonS3/latest/userguide/ObjectsinRequesterPaysBuckets.html

Because clickhouse doesn't sign them properly

**Expected behavior**

custom_headers working fine with aws s3.

### PR DESCRIPTION
**Pull request by arthurpassos**
Closes https://github.com/ClickHouse/ClickHouse/issues/27140

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Signs all `x-amz-*` headers as required by AWS S3 docs.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by wiledusc — APPROVED** [review]

File: src/IO/S3/Requests.h [hunk 1]
```diff
@@ -67,10 +67,36 @@ class ExtendedRequest : public BaseRequest
         api_mode = api_mode_;
     }
 
+    void AddAdditionalCustomHeader(const Aws::String & header_name, const Aws::String & header_value) const
+    {
+        additional_custom_headers.emplace(header_name, header_value);
+    }
+
+    /*
+     * Hackish as F!
```
Comment (wiledusc) [inline src/IO/S3/Requests.h:76]:
```suggestion
```

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 5d7d3c6935a608beb8ff1cb4b7b36845e18b7134 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/56058/5d7d3c6935a608beb8ff1cb4b7b36845e18b7134/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6658925401">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/56058/5d7d3c6935a608beb8ff1cb4b7b36845e18b7134/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
@arthurpassos `test_delayed_replica_failover/test.py::test` what is that?

**Comment by arthurpassos** [comment]
> @arthurpassos `test_delayed_replica_failover/test.py::test` what is that?

It is also failing on https://github.com/ClickHouse/ClickHouse/pull/56423

**Comment by arthurpassos** [comment]
@alexey-milovidov 

test_delayed_replica_failover - https://github.com/ClickHouse/ClickHouse/pull/56429

File: src/IO/S3/Client.cpp [hunk 2]
```diff
@@ -267,17 +267,32 @@ void Client::setKMSHeaders(RequestType & request) const
     }
 }
 
+
+template <typename RequestType>
```
Comment (alexey-milovidov) [inline src/IO/S3/Client.cpp:271]: Rewrite without templates.

`SetAdditionalCustomHeaderValue` is a virtual method.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by alexey-milovidov** [comment]
Squash it to get rid of commit messages `tmp` and `very hackish approach..`

File: src/IO/S3/Client.cpp (hunk 2 above)
Comment (alexey-milovidov) [inline src/IO/S3/Client.cpp:271]: The function has to be put into an anonymous namespace.

Comment (arthurpassos) [reply src/IO/S3/Client.cpp]: done

**Comment by arthurpassos** [comment]
> Squash it to get rid of commit messages `tmp` and `very hackish approach..`

Git history is a bit too complex to actually do it because of the `merges`. I can start a new branch & a new PR if that is mandatory. Wdyt?

**Comment by arthurpassos** [comment]
> > Squash it to get rid of commit messages `tmp` and `very hackish approach..`
> 
> @alexey-milovidov Git history is a bit too complex to actually do it because of the `merges`. I can start a new branch & a new PR if that is mandatory. Wdyt?

@alexey-milovidov in case you are interested https://github.com/ClickHouse/ClickHouse/pull/57001. CI needs to be enabled

**Pull request closed without merging**